### DISCUSSION TITLE
New compiler errors with TypeScript 4.5

### DISCUSSION THREAD

**Discussion by andrewbranch** [react-hook-form/react-hook-form#6943] [General] [answered]
Hello from the TypeScript team 👋 

We’ve noticed some compiler errors turn up in your codebase with TypeScript 4.5 (about to go into RC, stable in ~two weeks) that were not present in 4.4: https://github.com/microsoft/TypeScript/issues/46610

We looked into a few of these briefly today, and they look like _probably_ correct/accurate errors, but none of us are particularly familiar with this library, so we wanted to give you a heads up so you could have a head start on updating your code for 4.5 compatibility, or let us know if something looks like a bug.

I suspect that https://github.com/microsoft/TypeScript/pull/46429 is related, so you may want to review the changes to the rules of conditional type assignability mentioned in that PR description.

Thanks!

**Comment by bluebill1049** [comment, marked as answer]
thanks @andrewbranch cc @barrymay & @kotarella1110

**Reply by bluebill1049** [reply]
More context: I did encounter an issue when I was trying out the `4.5 beta`, worth for us to look into it why and `Type instantiation is excessively deep and possibly infinite` as far as I know from @barrymay mentioned ts 4.5 was meant to improve that: https://github.com/react-hook-form/react-hook-form/pull/6753#issuecomment-939222107 however it seems to do the opposite for us, probably not following the rule correctly.

**Reply by andrewbranch** [reply]
We’ve actually had some pretty significant changes since October 8 that eliminate some errors like that, although in your case it seems we simultaneously added some assignability errors. Definitely make sure to try out the latest nightly (or the RC in a couple days) as some things have definitely shifted around.

**Reply by bluebill1049** [reply]
thanks for the heads up will test out 4.5 today

**Reply by barrymay** [reply]
Hi @andrewbranch and @bluebill1049 I've been playing around with this for the past 30 and I can tell this may be quite tricky to work out, possibly because these are cases where tsc and the IDE might have "given up" previously and returned any's (particularly with nesting of generics) and allowed us to have generic types that may or may not have been kosher in the end.

I do have some ideas that I want to play around with, but I have a question for you @andrewbranch - is there a good best practice for unit testing our type definitions, particularly in the generic space?

**Reply by andrewbranch** [reply]
> tsc and the IDE might have "given up" previously and returned any's

Yeah, this was my best guess about what's going on.

With regard to testing types, several of the new failures in 4.5 occur in your test suite, which suggests that you already have the basis for what could be seen as integration or end-to-end testing of types, which I think is the most important thing, because it’s the same breaks users will see when they try to upgrade their TypeScript version. We just happen to periodically test your codebase against our nightly compiler builds because it’s an open-source collection of complex conditional type usage, and those real-world sources often turn up breaks that our internal test suite doesn’t expose. If you wanted to monitor for those kinds of breaks yourself, you could set up a scheduled CI build with `typescript@next` and catch these things even sooner.

We typically think systems like this, where you keep an eye on compiler errors in your source code and your test suites, are usually enough for most libraries. But you do have some super complicated types, so maybe unit testing would be beneficial in your case. I usually recommend looking at https://github.com/SamVerschueren/tsd for this.

**Reply by bluebill1049** [reply]
Thanks for the useful information and tips @andrewbranch. we have fixed all the issues related to `4.5.0-dev`. We will keep an eye on the RC release, and make sure everything still works with the latest.

**Reply by barrymay** [reply]
Thanks Andrew!  I had heard about tsd (and sadly forgot it) so thanks for the suggestion and the vote,  I'll check it out.  I agree normal unit testing is find for most libs but we've created some complex ones so it would be great to get these clarified for many reasons :)

@bluebill1049 has helped stabilize the library so we should technically be good to go for day one (using as clauses)  but we definitely want to sort this out with the stronger support in 4.5.  

It goes without saying, but **thank you** for all you guys do - TypeScript makes a big different to our sanity to ensure we're getting this right, while giving us the flexibility and awareness to "shoot ourselves in the foot" if we need to get by for a bit while we get it right.

**Reply by andrewbranch** [reply]
Awesome! I’m relieved to hear you were able to ~~fix~~ work around the issues. (Hopefully proper fixes aren’t too much work 😅)  Cheers! 👋
